datamol/viz/_substructure.py

```diff
@@ -0,0 +1,104 @@
+import itertools
+
+from typing import Union
+from typing import List
+
+import datamol as dm
+
+from ._viz import to_image
+
+
+def find_bonds_from_atom_list(mol: dm.Mol, atom_indices: List[List[int]]):
+    """Return a list of list of existing bond indices between a list
+    of list of atom indices.
+
+    Args:
+        mol: A molecule.
+        atom_indices: A list of list of atom indices.
+    """
+
+    bonds = []
+
+    for atom_indices_single in atom_indices:
+
+        bonds_single = []
+
+        # Test all possible combinations for existing bonds
+        for a1_idx, a2_idx in itertools.combinations(atom_indices_single, 2):
+            bond = mol.GetBondBetweenAtoms(a1_idx, a2_idx)
+            if bond is not None:
+                bonds_single.append(bond.GetIdx())
+
+        bonds.append(bonds_single)
+
+    return bonds
+
+
+def match_substructure(
+    mols: Union[List[dm.Mol], dm.Mol],
+    patterns: Union[List[dm.Mol], dm.Mol],
+    highlight_bonds: bool = True,
+    **kwargs,
+):
+    """Generate an image of molecule(s) with substructure matches for a given
+    pattern or substructure.
+
+    Args:
+        mols: One or more molecules.
+        patterns: One or more patterns.
+        highlight_bonds: Whether to also highlight the bonds matching the patterns.
+        kwargs: Other kwargs passed to `dm.viz.to_image`.
+    """
+
+    # NOTE(hadim): `MolsToGridImage` used in `to_image` can't use a list of list of indices
+    # for every molecules so it's not really possible to have different colors for different
+    # matches in the same molecules.
+    # In the future, we will implement our custom `MolsToGridImage` in order to have more controls
+    # on the colors used.
+    # For the same reason, we don't bother about colors here.
+
+    if isinstance(mols, dm.Mol):
+        mols = [mols]
+
+    if isinstance(patterns, dm.Mol):
+        patterns = [patterns]
+
+    # Always copy mols and patterns
+    mols = [dm.copy_mol(mol) for mol in mols]
+    patterns = [dm.copy_mol(mol) for mol in patterns]
+
+    all_atom_indices = []
+    all_bond_indices = []
+
+    for mol in mols:
+
+        atom_indices = []
+        bond_indices = []
+
+        for pattern in patterns:
+
+            atom_indices += list(mol.GetSubstructMatches(pattern, uniquify=True))
+
+            if highlight_bonds:
+                bond_indices += dm.viz.find_bonds_from_atom_list(mol, atom_indices)
```

Comment:
atom_indices would be sorted by the pattern atom indices.

It's actually both faster and more accurate to use that to get the bond_indices that matches (just iterate through the bond of pattern, get the two atoms (they will always be in the match), then get their equivalent in the molecule using the `matched indices` to deduce the bond in the molecule.

For why it's not precise, consider the below

![image](https://user-images.githubusercontent.com/5290110/144654249-f3ad89b6-8351-435f-af71-ff4d21b60218.png)

The wanted match should not involve the `N=N` bonds (as highlighted in the image), but since bonds between all combination of atoms in the atom list are considered, here we get a highlight that would cover that bond too.
Another example:
```
mol = dm.to_mol("C1CC2CCCC12")
query = dm.to_mol("C1CCCCCC1")
```
The bridge bond should not be highlighted (note that RDKIT default bond highlighting also get it very wrong by doing more or less the same thing as above). 

Using the function `atom_list_to_bond` above for each list of atom match should in theory fix the issues, but iterating through pattern bond is 100% correct without any surprising behaviour.

Reply:
Thanks, this way is clearly more elegant and faster indeed.

Reply:
I have added `dm.substructure_matching_bonds` which does a substructure match but also return bond matching bond indices in addition to matching atom indices. The logic follows the ones you were thinking of: use the order of `GetSubstructMatches` matches and an atom mapping then iterate on `query` to get the bond indices.

This is what `dm.viz.match_substructure` now use. I have no idea how robust this is but so far on the couple of examples I have tested it works well (I am sure we can find edge cases that will fail).